### PR TITLE
123 copy methods

### LINKED ISSUE #123: Copy/clone methods
A request for methods that clone or copy objects, coming from OPIL user:
https://github.com/SD2E/opil/issues/50

### PR DESCRIPTION
**Pull request by bbartley**
- Factored out dictionary of builder methods from `Document` class
- Created global `BUILDER_REGISTRY` and placed in object.py to avoid circular references
- Cloning an object from one `Document` to another is supported
- Resolve #123
- Resolve https://github.com/SD2E/opil/issues/50

New issues should be created for the following:
- Cloning an object within a single `Document` is *not* supported (need to figure out how to deal with URI collisions)
- Copying into new namespaces not currently supported either

### CODE REVIEW

**Comment by bbartley** [comment]
Since the global variable BUILDER_REGISTER is an offensive code smell, we might add it as a static variable in the `SBOLObject` class in order to better encapsulate it.

**Comment by bbartley** [comment]
Also, if the goal is to eventually deprecate `SBOLObject` we can move the `BUILDER_REGISTER` and `copy` methods to `Identified`.  I think that should be fairly easy.

File: test/test_object.py [hunk 1]
```diff
@@ -18,6 +18,65 @@ def test_trailing_slash(self):
...
+        doc.add(root)
+        doc.add(sub1)
+        doc.add(sub2)
+
+        doc2 = sbol3.Document()
+        root_copy = root.copy(target_doc=doc2)
+        self.assertEqual([sc.identity for sc in root.features],
+                         [sc.identity for sc in root_copy.features])
+
+    # def test_import_object_into_new_namespace(self):
```
Comment (tcmitchell) [inline test/test_object.py:46]: Can this commented out test method either be documented or removed?

If it is for a future capability, let's put in some comments explaining that so we know why it's there. If it served as a model for other tests and is not expected to ever be a pySBOL3 test then let's remove it.

File: sbol3/document.py [hunk 2]
```diff
@@ -11,6 +10,7 @@
 from rdflib_jsonld import parser as jsonld_parser
 
 from . import *
+from .object import BUILDER_REGISTER
```
Comment (tcmitchell) [inline sbol3/document.py:13]: This import implies that `BUILDER_REGISTER` is not available by importing sbol3. How will custom classes get access to `BUILDER_REGISTER`?

Did I miss something?

File: sbol3/document.py [hunk 3]
```diff
@@ -2,7 +2,6 @@
 import logging
 import warnings
 from typing import Dict, Callable, List, Optional
-
```
Comment (tcmitchell) [inline sbol3/document.py:5]: We'd like to keep this separation between imports from core Python modules and imports from third party modules.

File: sbol3/object.py [hunk 4]
```diff
@@ -3,7 +3,7 @@
 from collections import defaultdict
 from typing import Optional
 from urllib.parse import urlparse
-
+from typing import Dict, Callable
```
Comment (tcmitchell) [inline sbol3/object.py:6]: Can this be cleaned up so the `typing` imports are all on one line, and there is a blank line between core Python imports and the local import (`from . import *`)?

File: test/test_object.py (hunk 1 above)
Comment (bbartley) [reply test/test_object.py]: OK, I'll remove it for now.  That's a test case ported over from pySBOL2 that deals with namespace replacement.

File: sbol3/document.py (hunk 3 above)
Comment (bbartley) [reply sbol3/document.py:5]: OK, this is fixed now.  I never understood before what the rhyme or reason to how the imports were structured.

File: sbol3/object.py (hunk 4 above)
Comment (bbartley) [reply sbol3/object.py]: fixed

File: sbol3/document.py (hunk 2 above)
Comment (bbartley) [reply sbol3/document.py:13]: Custom classes (and the core classes which also register their builder) are registered using the `Document.register_builder` static method.  So no change.

`BUILDER_REGISTER` need only be accessible to the `Document` class.  See this line:
https://github.com/SynBioDex/pySBOL3/blob/123-copy_methods/sbol3/document.py#L37

**Review by tcmitchell — APPROVED** [review]

**Pull request merged**